lib/service/rpmsg/rpc/rpmsg_rpc_server.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <metal/mutex.h>
+#include <metal/spinlock.h>
+#include <metal/utilities.h>
+#include <openamp/open_amp.h>
+#include <openamp/rpmsg_rpc_client_server.h>
+#include <string.h>
+#include <fcntl.h>
+
+#define LPRINTF(format, ...) metal_log(METAL_LOG_ERROR, format, ##__VA_ARGS__)
+
+int rpmsg_rpc_server_init(struct rpmsg_rpc_svr *rpcs, struct rpmsg_device *rdev,
+			  const char *ept_name,
+			  const struct rpmsg_rpc_services *services, int len,
+			  rpmsg_ns_unbind_cb rpmsg_service_server_unbind)
+{
+	int ret;
+
+	rpcs->services = services;
+	rpcs->n_services = len;
+
+	ret = rpmsg_create_ept(&rpcs->ept, rdev, ept_name,
+			       RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
+			       rpmsg_endpoint_server_cb,
+			       rpmsg_service_server_unbind);
+	if (ret)
+		return ret;
+
+	return ret;
+}
+
+static const struct rpmsg_rpc_services *find_service(struct rpmsg_rpc_svr *rpcs,
+						     unsigned int id)
+{
+	for (unsigned int i = 0; i < rpcs->n_services; i++) {
+		if ((&rpcs->services[i])->id == id) {
+			return &rpcs->services[i];
+		}
+	}
+	return NULL;
+}
+
+int rpmsg_endpoint_server_cb(struct rpmsg_endpoint *ept, void *data,
+			     size_t len,
+			     uint32_t src, void *priv)
+{
+	unsigned char buf[MAX_BUF_LEN];
+	unsigned int id;
+	const struct rpmsg_rpc_services *service;
+	struct rpmsg_rpc_svr *rpcs;
+	(void)priv;
+	(void)src;
+
+	if (len > MAX_BUF_LEN)
+		return -EINVAL;
+
+	rpcs = metal_container_of(ept, struct rpmsg_rpc_svr, ept);
+
+	memcpy(buf, data, len);
+	id = *buf;
+	service = find_service(rpcs, id);
+
+	if (service) {
+		service->respcb.cb_function(buf, rpcs);
```

Comment:
The callback returns a status. Status should be tested and an error message printed